server/src/analyser.ts

```diff
@@ -275,6 +275,131 @@ export default class Analyzer {
+      ) {
+        const functionBody = parent.lastChild
+        TreeSitterUtil.forEach(functionBody, (n) => {
+          if (
+            (originalDeclaration && !continueSearching) ||
+            n.startPosition.row > boundary ||
+            n.type === 'function_definition' ||
+            n.type === 'subshell'
+          ) {
+            return false
+          }
+
+          if (n.type === 'declaration_command') {
+            const isLocal = ['local', 'declare', 'typeset'].includes(
+              n.firstChild?.text ?? '',
+            )
+            const firstInstance = n.descendantsOfType('variable_name').at(0)
+            // Check for var="$var" cases
+            const instanceInExpression =
+              n.endPosition.row >= position.line &&
+              firstInstance &&
+              (firstInstance.endPosition.column < position.character ||
+                firstInstance.endPosition.row < position.line)
+            if (isLocal && firstInstance?.text === word && !instanceInExpression) {
+              originalDeclaration = firstInstance
+              continueSearching = false
+            }
+
+            return false
+          }
+
+          return true
+        })
+      } else if (parent.type === 'subshell') {
+        TreeSitterUtil.forEach(parent, (n) => {
+          if (
+            (originalDeclaration && !continueSearching) ||
+            n.startPosition.row > boundary ||
+            (n.type === 'subshell' && !parent?.equals(n))
+          ) {
+            return false
+          }
+
+          let node: Parser.SyntaxNode | null | undefined
+
+          if (
+            type === 'variable' &&
+            (n.type === 'declaration_command' ||
+              n.type === 'variable_assignment' ||
+              n.type === 'for_statement' ||
+              (n.type === 'command' && n.text.includes(':=')))
+          ) {
+            node = n.descendantsOfType('variable_name').at(0)
+          } else if (type === 'function' && n.type === 'function_definition') {
+            node = n.firstNamedChild
+          }
+
+          // Check for var="$var" cases
+          const instanceInExpression =
+            type === 'variable' &&
+            n.type !== 'for_statement' &&
+            n.endPosition.row >= position.line &&
+            node &&
+            (node.endPosition.column < position.character ||
+              node.endPosition.row < position.line)
+          if (node?.text === word && !instanceInExpression) {
+            originalDeclaration = node
+            continueSearching = n.type === 'command'
+          }
+
+          if (n.type === 'function_definition') {
+            return false
+          }
+
+          return true
+        })
+      }
+
+      if (originalDeclaration && !continueSearching) {
+        break
+      }
+
+      boundary = parent.startPosition.row
+      parent = this.findParentScopeNode(uri, parent.startPosition, parent.endPosition)
+    }
+
+    // TODO: Handle global definitions
```

Comment:
Do you mean declarations across files? `getGlobalDeclarations` should help with this.